### PR TITLE
Context Menu rendering

### LINKED ISSUE #2967: [HOLD for payment 23 June] Web - Chat - Adjust modal rendering
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Right clicking on a chat message should open the context menu (i.e. right click menu). Right clicking on a second chat message (without closing the context menu on the first chat) should open the custom context menu. Right clicking outside of the currently open menu should close that menu. 

## Actual Result:
Right clicking on a chat message opens the context menu (intended result). Right clicking on a second chat message (without closing the context menu on the first chat) does not open any contact menu. Instead, browser default context menu shows (web), or nothing happens (desktop app). 

## Desired Outcome: 
We believe this issue occurs because the menu is rendered over a full screen background. We'd like to eliminate the full screen context menu background or otherwise rendered in a way that allows a user to click outside of the currently open menu. 

## Action Performed:
**On web browser:** 
1. Open expensify.cash in a web browser and log in 
2. Click on a chat in the left hand menu to show the messages 
3. Right click on a message, the context menu appears (intended result) 
4. Right click on a different message: browser default context menu is shown 

**On web app**
1. Open the expensify.cash desktop app and  log in 
2. Click on a chat in the left hand menu to show the messages 
3. Right click on a message, the context menu appears (intended result) 
4. Right click on a different message: nothing happens, context menu remains open on the first message you clicked in Step 

## Workaround:
N/A

## Platform:
Where is this issue occurring?

 Web ✔️
 iOS
 Android
 Desktop App ✔️
 Mobile Web

**Version Number:** 1.0.50-0
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/118547738-7fc29c00-b727-11eb-9482-a3a63947df69.mp4

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

### PR DESCRIPTION
**Pull request by parasharrajat**
### Fixed Issues

Fixes #2967

### Tests | QA Steps (Web | desktop)
1. Open Any conversation on E.cash.
2. Open the Context menu on any message.
3. Now try to open the context menu for another message. The previous message should close and a new one should open.
4. Open the context menu on any message.
5. Click somewhere outside the context menu. A context menu should close.

### Tests | QA Steps (Mobile | Mweb)

1. open Any conversation on E.cash
2. Hold press any message to open the context menu.
3. Click in the empty area should close the menu.
3. There should be no apparent changes and the context menu should behave as before.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [ ] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/24370807/121562741-5ef41a80-ca37-11eb-98f3-4b6cd684601e.mp4

#### Mobile Web

https://user-images.githubusercontent.com/24370807/121562414-0c1a6300-ca37-11eb-980a-36ad53d42ffd.mp4

#### Desktop

https://user-images.githubusercontent.com/24370807/121561838-7e3e7800-ca36-11eb-978e-ff15e66fb7e5.mp4
#### Android

https://user-images.githubusercontent.com/24370807/121562428-0f155380-ca37-11eb-9209-7f27ecb8095c.mp4

### CODE REVIEW

**Review by Dal-Papa — DISMISSED** [review]
This LGTM but I'd rather have @marcaaron or @Julesssss review the code than me.

File: src/styles/getModalStyles.js
Comment (Beamanator) [inline src/styles/getModalStyles.js:144]: I'm interested to know why we've been doing this in this file a lot, since I don't think it's necessary - I asked in [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1623398725207000)

**Comment by Julesssss** [comment]
Sorry, I'm unable to take additional reviews today.

**Comment by parasharrajat** [comment]
@marcaaron  any thoughts. Do you see any gotchas to component composition here?

**Comment by marcaaron** [comment]
>Do you see any gotchas to component composition here?

Hmm sorry, can you ask that in another way? I'm not so sure what you meant.

**Comment by parasharrajat** [comment]
Sorry, Could you please review it, and do you think is there any better way of reusing the popover as I feel we have added a couple of wrappers around the baseModal.

**Comment by marcaaron** [comment]
Perhaps request in the Slack channel to see if someone else can provide a second opinion?

**Comment by parasharrajat** [comment]
@Beamanator All yours. Thanks.

File: src/CONST.js
Comment (roryabraham) [inline src/CONST.js:87]: Since the only difference between these two modal types is the value of a couple props, `hasBackdrop` and `coverScreen`, I don't think we need a new modal type here. Just override those props in the places we need to for now. Better yet, just make them both false for the `Popover` modal type by default, which I think makes the most sense?

File: src/components/ContextMenuPopover/index.js
Comment (roryabraham) [inline src/components/ContextMenuPopover/index.js:8]: Why creating a portal? Some `zIndex` business? Can we leave a comment? Should we just make this change directly in `Popover` instead of creating a new component?

File: src/components/Modal/index.js
Comment (roryabraham) [inline src/components/Modal/index.js:1]: We weren't using `memo` before, so why start using `PureComponent` now?

File: src/styles/getModalStyles.js
Comment (roryabraham) [inline src/styles/getModalStyles.js:136]: As noted above, I don't think that this modal type is necessary. But if we _did_ keep it, then we could simplify this and remove a lot of duplicate code:

```javascript
case CONST.MODAL.MODAL_TYPE.POPOVER:
case CONST.MODAL.MODAL_TYPE.CONTEXT_MENU:
    // all the default popover stuff
    
    if (type === CONST.MODAL.MODAL_TYPE.CONTEXT_MENU) {
        // This is the only difference between POPOVER and CONTEXT_MENU
        coverScreen = false;
        hasBackdrop = false;
    }
    break;
```

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/components/ContextMenuPopover/index.native.js
Comment (parasharrajat) [reply src/components/ContextMenuPopover/index.native.js:23]: Yeah. I saw the same thing on POPover so I will remove it there as well.

File: src/components/Modal/index.js
Comment (parasharrajat) [reply src/components/Modal/index.js:1]: Yeah, it can be removed. `baseModal` is already pure.

File: src/styles/getModalStyles.js
Comment (parasharrajat) [reply src/styles/getModalStyles.js:136]: Makes a lot of sense, I was just not able to think to minimize it but I knew It can be.

File: src/CONST.js
Comment (parasharrajat) [reply src/CONST.js:87]: I agree, working on merging the Popover and ContextMenu into Popover

File: src/components/ContextMenuPopover/index.js
Comment (parasharrajat) [reply src/components/ContextMenuPopover/index.js:8]: Z-index does not work here. z-index is always relative to the parent. If its parent sibling has a higher z-index that will cover the Menu.

**Comment by parasharrajat** [comment]
Full refactor. Thanks.

**Review by Beamanator — DISMISSED** [review]
Looks awesome, tested and works well 👍  Will let Rory check it before merging

**Review by roryabraham — DISMISSED** [review]
Nice, looks great!

File: src/components/Popover/index.js
Comment (Beamanator) [inline src/components/Popover/index.js:17]: Hmm now that this prop is set conditionally below (from this PR: https://github.com/Expensify/Expensify.cash/pull/3558), should we do the same here?
```suggestion
                popoverAnchorPosition={props.isSmallScreenWidth ? undefined : props.anchorPosition}
```

Comment (parasharrajat) [reply src/components/Popover/index.js:17]: No as here this will never render on mobile. @beamanator

Comment (Beamanator) [reply src/components/Popover/index.js:17]: Ok because on mobile `props.isSmallScreenWidth` will always be `true` 👍

**Review by Beamanator — APPROVED** [review]
#LGTM!

**Comment by isagoico** [comment]
@parasharrajat We were still able to reproduce this issue https://github.com/Expensify/Expensify.cash/issues/3636 in mWeb.

**Comment by parasharrajat** [comment]
@isagoico Yeah there as accidentally bad merge of two PR which resulted in this behaviour. You are correct that other PR is going yo fix all the issues with context menu. I believe it will be merge today. So let's wait. Otherwise you can check it off, if that is a good option at the moment.

**Pull request merged**
